src/main/scala/org/zalando/zhewbacca/RequestValidator.scala

```diff
@@ -13,13 +13,14 @@ private[zhewbacca] object RequestValidator {
   def validate[A](scope: Scope, requestHeader: RequestHeader, authProvider: AuthProvider)(implicit ec: ExecutionContext): Future[Either[Result, TokenInfo]] = {
     authProvider.valid(OAuth2Token.from(requestHeader), scope).map {
       case AuthTokenValid(tokenInfo) => Right(tokenInfo)
-      case AuthTokenInvalid => Left(Results.Forbidden)
+      case AuthTokenInvalid => Left(Results.Unauthorized)
       case AuthTokenEmpty => Left(Results.Unauthorized)
+      case AuthTokenInsufficient => Left(Results.Forbidden)
     } recover {
       case NonFatal(e) =>
         logger.error(e.getMessage, e)
-        logger.debug("Request forbidden because of failure in Authentication Provider")
-        Left(Results.Forbidden)
+        logger.debug("Request unauthorized because of failure in Authentication Provider")
+        Left(Results.Unauthorized)
```

Comment:
I'm wondering if this should be represented as 502 or 503 as this is not a client's mistake but internal problem

Reply:
The access token provided is expired, revoked, malformed, or
         invalid for other reasons.  The resource SHOULD respond with
         the HTTP 401 (Unauthorized) status code.  The client MAY
         request a new access token and retry the protected resource
         request.


>>>  invalid for other reasons. 
i would go for 401

Reply:
well, in general 4xx is a client problem which means client is doing something wrong and should change the data before retrying 
in this case it is definitely a server problem, right? So this _should_ be something 5xx
In the passage you're referring to "invalid for other reasons" applies to the token, not in general and in the situation at hand we just can't tell if it is valid or not. And this is server's fault.